cmd/otel-allocator/README.md

```diff
@@ -43,6 +43,20 @@ spec:
           processors: []
           exporters: [logging]
 ```
+## Set up filterStrategy
+`filterStrategy` determines how to filter targets before allocating them among multiple collectors. `relabel-config` is the currently supported option. With `filterStrategy`, targets will be consolidated based on Prometheus configuration `relabel_config`. By default, filtering is disabled.
+
+It's important to consider setting `filterStrategy` if you have a scrape job with a discovery role for example, `endpoints` which could produce multiple targets per endpoint. If these targets lead to the same metric after a relabeling step and they are assigned to different collectors, the outcome may be undesirable because duplicate time series are created for the metric. 
```

Comment:
IIRC filter strategy doesn't solve this issue, it would still produce the same results as before (i.e. duplication of scrape targets). This would only be solved by changing the scrape configuration or the service/pod monitor.

Reply:
We have a metrics pipeline with number of jobs scraping EKS related metrics, see an example job below. Our pipeline setup is TA +  multi-collectors (min 3 with autoscaler). Each collector consists of a Prometheus receiver plus Prometheus remote write exporter.  We've seen lot of out of order sample errors reported from the pipeline, intermittently however persistently.  

```
- job_name: kubernetes-service-endpoints
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 1m
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: drop
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_node_name]
    separator: ;
    regex: (.*)
    target_label: node
    replacement: $1
    action: replace
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
...
```

After loads of debugging,  we discovered one of the causes. Two main factors contributed to the error: the scrape config from Prometheus and allocation of the discovered targets for the same eventual metric got split across different collector instances. 

Let's look at an example. With the role: `endpoints`,  for each endpoint address, a target is discovered per port.  For a service endpoint having three unique port names,  three targets 10.244.0.4:53, 10.244.0.4:53 and 10.244.0.4:9153 were discovered. We've seen  these 3 targets got split as following

```
collector #1: 10.244.0.4:53

collector #2: 10.244.0.4:53
collector #2: 10.244.0.4:9153
```
However,  after step below,  these three targets ended up with an identical label `instance=10.244.0.4:9153` for the eventual metric.  

```
- source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
```

We observed duplicated time series for the same metric were produced by these collectors with slightly shifted time stamps. Intermittently, the shifted time stamps triggered the out of order sample error if older samples were sent to PWR. 
  
You are right, we had experimented with changing the scrape configuration to get rid of the error. Considering the number of the jobs we have, the process is tedious and we're looking for alternatives. @Aneurysm9 suggested to look into `filterStrategy`.  Since then, we've been testing the fix in our pipeline.  Results so far, 

- We've seen good error reduction rate for multiple scrape jobs, job`kubernetes-service-endpoints` is one of them.
- We've only had moderate success of using this with scrape job `kubernetes-pods` (see config below) which we are still debugging. 
- We're also seeing the `out of order sample` errors reported with `target_info`.

Previously, we had reproduced the out of order sample error in a setup of single collector (no TA).  In that case,  this setting is not applicable. 

```- job_name: kubernetes-pods
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 1m
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
...
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: drop
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_phase]
    separator: ;
    regex: Pending|Succeeded|Failed|Completed
    replacement: $1
    action: drop
  kubernetes_sd_configs:
  - role: pod
    kubeconfig_file: ""
    follow_redirects: true
```

Reply:
Although tedious, editing your scrape config to prevent dual scrapes is the correct resolution here. You're most likely seeing out of order scrapes still because the filter strategy doesn't necessarily do the functionality you are expecting here (i.e. if any of the labels discovered for a pod / scrape config are different they will still result in two separate targets) In some cases you may get lucky that your scrape config removes the duplicates, but that's not a panacea.

Reply:
We made changes to our scrape config to introduce additional labels as well as collector_name, results have been looking stable so far, most of the errors were gone except `up`, `scraped_*` and `target_info` which we'll continue investigate.
@jaronoff97 @pavolloffay, thank you for your feedback.  I'll continue to evaluate other viable alternatives e.g.  logic from Prometheus to handle duplicated target ([code](https://github.com/prometheus/prometheus/blob/80b7f73d267a812b3689321554aec637b75f468d/scrape/scrape.go#L539)) for this. For now, I will close this PR.